transports/webrtc/src/connection.rs

```diff
@@ -0,0 +1,271 @@
+        return Poll::Pending;
+    }
+
+    fn poll_outbound(&self, cx: &mut Context<'_>) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner.lock().unwrap();
+        let peer_conn = self.peer_conn.clone();
+        let fut = inner.outbound_fut.get_or_insert(Box::pin(async move {
+            let peer_conn = peer_conn.lock().await;
+
+            // Create a datachannel with label 'data'
+            let data_channel = peer_conn
+                .create_data_channel("data", None)
+                .map_err(Error::WebRTC)
+                .await?;
+
+            trace!("Opening outbound substream {}", data_channel.id());
+
+            // No need to hold the lock during the DTLS handshake.
+            drop(peer_conn);
+
+            let (tx, rx) = oneshot::channel::<Arc<DetachedDataChannel>>();
+
+            // Wait until the data channel is opened and detach it.
+            register_data_channel_open_handler(data_channel, tx).await;
+
+            // Wait until data channel is opened and ready to use
+            match rx.await {
+                Ok(detached) => Ok(detached),
+                Err(e) => Err(Error::InternalError(e.to_string())),
+            }
+        }));
+
+        match ready!(fut.as_mut().poll(cx)) {
+            Ok(detached) => {
+                let mut ch = PollDataChannel::new(detached);
+                if let Some(cap) = inner.read_buf_cap {
+                    ch.set_read_buf_capacity(cap);
+                }
+
+                Poll::Ready(Ok(ch))
+            }
+            Err(e) => Poll::Ready(Err(e)),
+        }
+    }
+
+    fn poll_close(&self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        debug!("Closing connection");
+
+        let mut inner = self.inner.lock().unwrap();
+        let peer_conn = self.peer_conn.clone();
+        let fut = inner.close_fut.get_or_insert(Box::pin(async move {
+            let peer_conn = peer_conn.lock().await;
+            peer_conn.close().await.map_err(Error::WebRTC)
+        }));
+
+        match ready!(fut.as_mut().poll(cx)) {
+            Ok(()) => {
+                inner.incoming_data_channels_rx.close();
+                Poll::Ready(Ok(()))
+            }
+            Err(e) => Poll::Ready(Err(e)),
+        }
+    }
+}
+
+pub(crate) async fn register_data_channel_open_handler(
+    data_channel: Arc<RTCDataChannel>,
+    data_channel_tx: Sender<Arc<DetachedDataChannel>>,
+) {
+    data_channel
+        .on_open({
+            let data_channel = data_channel.clone();
+            Box::new(move || {
+                debug!(
+                    "Data channel '{}'-'{}' open",
+                    data_channel.label(),
+                    data_channel.id()
+                );
+
+                Box::pin(async move {
+                    let data_channel = data_channel.clone();
+                    match data_channel.detach().await {
```

Comment:
So I am reading the source to this and I am baffled about these API design choices.

From [this](https://docs.rs/webrtc/0.4.0/src/webrtc/data_channel/mod.rs.html#223) warning, I am taking that we actually need to call this from the `on_open` handler.

So. Many. Allocations. :(

Reply:
> From [this](https://docs.rs/webrtc/0.4.0/src/webrtc/data_channel/mod.rs.html#223) warning, I am taking that we actually need to call this from the on_open handler.

we're calling it from inside `on_open` handler. see line \#246